cpp/src/db/db_object.cu

```diff
@@ -0,0 +1,328 @@
+  void db_table<idx_t>::addEntry(db_pattern<idx_t>& pattern) {
+    if (!pattern.isAllConstants())
+      throw new std::invalid_argument("Can't add an entry that isn't all constants");
+    if (!pattern.getSize() != columns.size())
+      throw new std::invalid_argument("Can't add an entry that isn't the right size");
+    inputBuffer.push_back(pattern);
+  }
+
+  template<typename idx_t>
+  void db_table<idx_t>::rebuildIndices() {
+    for (size_t i = 0; i < columns.size(); i++) {
+      // Copy the column's data to a new array
+      idx_t size = columns[i]->size;
+      idx_t* tempColumn;
+      ALLOC_TRY(&tempColumn, sizeof(idx_t) * size, nullptr);
+      cudaMemcpy(tempColumn, columns[i]->data, sizeof(idx_t) * size, cudaMemcpyDefault);
+
+      // Construct an array of ascending integers
+      idx_t* indirection;
+      ALLOC_TRY(&indirection, sizeof(idx_t) * size, nullptr);
+      thrust::sequence(rmm::exec_policy(nullptr)->on(nullptr), indirection, indirection + size);
+
+      // Sort the arrays together
+      thrust::sort_by_key(rmm::exec_policy(nullptr)->on(nullptr),
+                          tempColumn,
+                          tempColumn + size,
+                          indirection);
+
+      // Compute offsets array based on sorted column
+      idx_t maxId;
+      cudaMemcpy(&maxId, tempColumn + size - 1, sizeof(idx_t), cudaMemcpyDefault);
+      idx_t *unique, *counts, *runCount;
+      ALLOC_TRY(&unique, (maxId + 1) * sizeof(idx_t), nullptr);
+      ALLOC_TRY(&counts, (maxId + 1) * sizeof(idx_t), nullptr);
+      ALLOC_TRY(&runCount, sizeof(idx_t), nullptr);
+      void* tmpStorage = nullptr;
+      size_t tmpBytes = 0;
+      cub::DeviceRunLengthEncode::Encode(tmpStorage,
+                                         tmpBytes,
+                                         tempColumn,
+                                         unique,
+                                         counts,
+                                         runCount,
+                                         size);
+      ALLOC_TRY(&tmpStorage, tmpBytes, nullptr);
+      cub::DeviceRunLengthEncode::Encode(tmpStorage,
+                                         tmpBytes,
+                                         tempColumn,
+                                         unique,
+                                         counts,
+                                         runCount,
+                                         size);
+      ALLOC_FREE_TRY(tmpStorage, nullptr);
+      idx_t runCount_h;
+      cudaMemcpy(&runCount_h, runCount, sizeof(idx_t), cudaMemcpyDefault);
+      idx_t* offsets;
+
+      // Allocating the new offsets array
+      ALLOC_TRY(&offsets, (maxId + 2) * sizeof(idx_t), nullptr);
+
+      // Filling values in offsets array from the encoded run lengths
+      int threadsPerBlock = 1024;
+      int numBlocks = (runCount_h + threadsPerBlock - 1) / threadsPerBlock;
+      offsetsKernel<<<numBlocks, threadsPerBlock>>>(runCount_h, unique, counts, offsets);
```

Comment:
May consider using thrust::permutation_iterator (as this code is mostly composed of thrust and cub calls).

Reply:
It doesn't seem like using thrust::permutation_iterator will simplify things much. I will leave this as is.